File: 28. Convert to doubly linked list inorder.py
class Node:
    def __init__(self,data): self.data=data; self.left=None; self.right=None

prev_node=None

def b_to_dll(root):
    global prev_node
    if not root: return None
    head=b_to_dll(root.left)
    if prev_node is None:
        head=root
    else:
        prev_node.right=root; root.left=prev_node
    prev_node=root
    b_to_dll(root.right)
    return head

def print_dll(head):
    cur=head
    while cur:
        print(cur.data, end=' '); cur=cur.right
    print()

if __name__=='__main__':
    root=Node(10); root.left=Node(12); root.right=Node(15); root.left.left=Node(25); root.left.right=Node(30)
    head=b_to_dll(root)
    print_dll(head)

File: 28. Convert to doubly linked list inorder.cpp
#include <bits/stdc++.h>
using namespace std;
struct Node
{
    int data;
    Node* left;
    Node* right;
    Node(int d) : data(d), left(nullptr), right(nullptr) {}
};
Node* prevNode = nullptr;
Node* bToDLL(Node* root)
{
    if (!root)
        return root;
    Node* head = bToDLL(root->left);
    if (prevNode == nullptr)
        head = root;
    else
    {
        prevNode->right = root;
        root->left = prevNode;
    }
    prevNode = root;
    bToDLL(root->right);
    return head;
}
void printDLL(Node* head)
{
    while (head)
    {
        cout << head->data << " ";
        head = head->right;
    }
    cout << endl;
}
int main()
{
    Node* root = new Node(10);
    root->left = new Node(12);
    root->right = new Node(15);
    root->left->left = new Node(25);
    root->left->right = new Node(30);
    Node* head = bToDLL(root);
    printDLL(head);
    return 0;
}
